### PR TITLE
compatibility with mariadb connector 3.2.8

### PR DESCRIPTION
**Pull request by hhaensel**
Currently usage of the latest mariadb odbc connector fails due to two reasons
- the termination character is not stripped off
- the max value for characters in a row is returned as 65535, which leads to a `Column length too big` error

Both root causes are addressed with this PR.

Superseded by #403, which includes the safe type-size fix from #401 and preserves valid trailing NUL data. The global NUL stripping in this PR was not merged.

### CODE REVIEW

File: src/dbinterface.jl [hunk 1]
```diff
@@ -340,7 +340,21 @@ function Cursor(stmt; iterate_rows::Bool=false, ignore_driver_row_count::Bool=fa
...
+                        count = length(raw_bytes)
+                        # strip trailing null bytes if it's a character/string data type
+                        # (Drivers often pad with \0, but Julia strings shouldn't embed them)
+                        if ctype == API.SQL_C_CHAR || ctype == API.SQL_C_WCHAR
+                            while count > 0 && raw_bytes[count] == 0x00
+                                count -= 1
+                            end
+                        end
+
+                        A[j] = jlcast(Base.nonmissingtype(T), view(raw_bytes, 1:count))
```
Comment (quinnj) [inline src/dbinterface.jl]: Could we move this trimming into the shared string conversion path? This only covers the columnar fetch path; row iteration still calls `jlcast(String, data[1:b.totallen])`, so `iterate_rows=true`/long-text fetches can still return strings with trailing `\0`. [findings by codex; reviewed by quinnj]

File: src/load.jl [hunk 2]
```diff
@@ -40,7 +40,9 @@ checkdupnames(names) = length(unique(map(x->lowercase(String(x)), names))) == le
 function createtable(conn::Connection, nm::AbstractString, sch::Tables.Schema; debug::Bool=false, quoteidentifiers::Bool=true, createtableclause::AbstractString="CREATE TABLE", columnsuffix=Dict())
     names = sch.names
     checkdupnames(names)
-    types = [sqltype(conn, T) for T in sch.types]
+    # prevent row too large error by limiting column size for long text types
+    # as some drivers (like mariadb-connectors-odbc 3.2.8) will return 65535
+    types = [replace(sqltype(conn, T), "65535" => "255") for T in sch.types]
```
Comment (quinnj) [inline src/load.jl]: Could this be narrowed to the specific MariaDB text type case? Replacing every `65535` in the SQL type string can silently shrink unrelated columns, e.g. `VARBINARY(65535)` or `DECIMAL(65535,6)`, to 255. [findings by codex; reviewed by quinnj]

File: src/dbinterface.jl (hunk 1 above)
Comment (hhaensel) [reply src/dbinterface.jl]: moved to jlcast

File: src/load.jl (hunk 2 above)
Comment (hhaensel) [reply src/load.jl]: VARBINARY(65535) is also not possible, so that also needs to be reduced. I searched a bit how to determine which connector is selected, but I couldn't find a way to distinguish between mariadb 3.1 and 3.2.
Maybe a more reasonable approach would be to pass a dict with type => size correlation or name => size correlation?

Comment (hhaensel) [reply src/load.jl]: along these lines ...
```julia
function createtable(
    conn::Connection, 
    nm::AbstractString, 
    sch::Tables.Schema; 
    debug::Bool=false, 
    quoteidentifiers::Bool=true, 
    createtableclause::AbstractString="CREATE TABLE", 
    columnsuffix=Dict(),
    columntypes=Dict()
)
    names = sch.names
    checkdupnames(names)
    types = [haskey(columntypes, names[i]) ? 
             string(columntypes[names[i]]) : 
             sqltype(conn, sch.types[i]) 
             for i = 1:length(names)]

    columns = (string(quoteidentifiers ? quoteid(conn, String(names[i])) : names[i], ' ', types[i], ' ', get(columnsuffix, names[i], "")) for i = 1:length(names))
    debug && @info "executing create table statement: `$createtableclause $nm ($(join(columns, ", ")))`"
    return DBInterface.execute(conn, "$createtableclause $nm ($(join(columns, ", ")))")
end
```
or these
```julia
function sqltype(conn, T, columnsize::Union{Integer, Nothing} = nothing)
    if isempty(conn.types)
        types = Tables.columntable(Cursor(API.gettypes(conn.dbc)))
        conn.alltypes = types
        i = findfirst(==(API.SQL_VARCHAR), types.DATA_TYPE)
        if i === nothing
            defaultT = "VARCHAR(255)"
        else
            cs = something(columnsize, types.COLUMN_SIZE[i])
            defaultT = types.TYPE_NAME[i] * "($cs)"
        end
        for jlT in BINDTYPES
            _, sqlT = bindtypes(jlT)
            i = findfirst(==(sqlT), types.DATA_TYPE)
            nm = i !== nothing ? types.TYPE_NAME[i] : defaultT
            if i !== nothing && types.CREATE_PARAMS[i] !== missing && nm != "DOUBLE" && nm != "FLOAT"
                cs = something(columnsize, types.COLUMN_SIZE[i])
                nm *= occursin(',', types.CREATE_PARAMS[i]) ? "($(typeprecision(jlT)),$(typescale(jlT)))" : "($cs)"
            end
            conn.types[jlT] = nm
        end
    end
    return conn.types[T]
end

checkdupnames(names) = length(unique(map(x->lowercase(String(x)), names))) == length(names) || error("duplicate case-insensitive column names detected; sqlite doesn't allow duplicate column names and treats them case insensitive")

function createtable(
    conn::Connection, 
    nm::AbstractString, 
    sch::Tables.Schema; 
    debug::Bool=false, 
    quoteidentifiers::Bool=true, 
    createtableclause::AbstractString="CREATE TABLE", 
    columnsuffix=Dict(),
    columnsizes=Dict()
)
    names = sch.names
    checkdupnames(names)
    types = [sqltype(conn, T, get(columnsizes, names[i], nothing)) for (i, T) in enumerate(sch.types)]
    columns = (string(quoteidentifiers ? quoteid(conn, String(names[i])) : names[i], ' ', types[i], ' ', get(columnsuffix, names[i], "")) for i = 1:length(names))
    debug && @info "executing create table statement: `$createtableclause $nm ($(join(columns, ", ")))`"
    return DBInterface.execute(conn, "$createtableclause $nm ($(join(columns, ", ")))")
end
```

Comment (hhaensel) [reply src/load.jl]: Or rather provide a conntypes to be merged with conn.types
Dict{Type, String} with 19 entries:
  Dec64         => "DECIMAL(16,6)"
  Vector{UInt8} => "VARBINARY(65535)"
  Float32       => "VARCHAR(65535)"
  Int64         => "BIGINT"
  Dec128        => "DECIMAL(35,6)"
  Bool          => "BIT"
  Time          => "TIME"
  UInt16        => "SMALLINT"
  Int8          => "TINYINT"
  UUID          => "VARCHAR(65535)"
  Date          => "DATE"
  UInt32        => "INTEGER"
  DateTime      => "DATETIME"
  Int32         => "INTEGER"
  UInt64        => "BIGINT"
  Float64       => "DOUBLE"
  String        => "VARCHAR(65535)"
  Int16         => "SMALLINT"
  UInt8         => "TINYINT"

Comment (hhaensel) [reply src/load.jl]: After some nightly reflection I think none of these proposals is a good one.
We should rather provide better type detection in `sqltype()`. I'll come up with a proposal.

Comment (hhaensel) [reply src/load.jl]: @quinnj what do you think about this approach. It tests whether default sizes have the typical excessive values of the mariadb driver 3.2 that don't make sense for default types.

```julia
function sqltype(conn, T)
    if isempty(conn.types)
        types = Tables.columntable(Cursor(API.gettypes(conn.dbc)))
        conn.alltypes = types
        i = findfirst(==(API.SQL_VARCHAR), types.DATA_TYPE)
        if i === nothing
            defaultT = "VARCHAR(255)"
        else
            # Safeguard: Any driver returning exactly the maximum tiny/medium text range 
            # for default VARCHAR triggers a standard 255 fallback, e.g. mariadb version 3.2
            colsize = types.COLUMN_SIZE[i]
            (colsize == 65535 || types.COLUMN_SIZE[i] == 16777215) && (colsize = 255)
            defaultT = types.TYPE_NAME[i] * "($colsize)"
        end
        for jlT in BINDTYPES
            _, sqlT = bindtypes(jlT)
            i = findfirst(==(sqlT), types.DATA_TYPE)
            nm = i !== nothing ? types.TYPE_NAME[i] : defaultT
            if i !== nothing && types.CREATE_PARAMS[i] !== missing && nm != "DOUBLE" && nm != "FLOAT"
                
                colsize = types.COLUMN_SIZE[i]
                # Step down the explicit type bindings if they hit MariaDB max sizes
                if (sqlT in (API.SQL_VARCHAR, API.SQL_VARBINARY, API.SQL_WVARCHAR)) && 
                   (colsize == 65535 || colsize == 16777215)
                    colsize = 255
                end
                
                nm *= occursin(',', types.CREATE_PARAMS[i]) ? "($(typeprecision(jlT)),$(typescale(jlT)))" : "($(colsize))"
            end
            conn.types[jlT] = nm
        end
    end
    return conn.types[T]
end
```

Comment (hhaensel) [reply src/load.jl]: Moreover, the README for testing is outdated, e.g. we need a mariadb for testing, an mysql will throw upon
```julia
result = DBInterface.execute(
    conn, "SELECT a, b, a/b AS c FROM (VALUES (2,1),(1,0),(2,1)) AS t(a,b)"
) |> columntable
```
and setting TRAVIS_DIR is deprecated, if I'm right.
Should I update that section in this PR?

Comment (quinnj) [reply src/load.jl]: Thanks, this makes sense. I agree that adding `columntypes`/`columnsizes`/`conntypes` feels too broad for this PR. The `sqltype()` approach is the direction I’d prefer: normalize the suspicious MariaDB max sizes while interpreting the driver metadata, and only for the variable string/binary SQL types, rather than doing a blanket string `replace` in `createtable`.

Could you update the branch with that approach and remove the `replace(sqltype(...), "65535" => "255")` call? I’d keep the README testing cleanup separate from this compatibility fix.

Comment (hhaensel) [reply src/load.jl]: @quinnj done

**Comment by hhaensel** [comment]
@quinnj anything to be done to get this merged?

**Comment by quinnj** [comment]
Not ready to merge at `7e57b9b`:

- `src/utils.jl:344-350`: stripping all trailing NUL bytes in the shared conversion loses stored data. I ran the exact function on Julia 1.12.6: `jlcast(String, UInt8[0x61, 0x00])` returns `"a"`, and `[0x00]` becomes empty. Both fetch paths already slice by the reported length. Please identify why 3.2.8 includes the terminator and correct that length handling (or use a narrowly scoped driver workaround), while preserving payload NULs. [ODBC returns the actual data byte length](https://learn.microsoft.com/en-us/sql/odbc/reference/develop-app/using-length-and-indicator-values).
- Add regression tests for both fetch modes, long text, empty strings, trailing payload NULs, and the type-size fallback. CI and the Linux test driver path still pin Connector 3.1.20, so the green Julia jobs do not validate the reported 3.2.8 issue. Please exercise 3.2.8 while retaining older-driver coverage. Both Codecov checks also currently fail.

I have not pushed a speculative change: the driver-specific cause still needs a reproducer before a safe fix can be validated.

**Comment by quinnj** [comment]
Root cause found, thanks for the report. MariaDB Connector/ODBC changed the `COLUMN_SIZE` it reports for `VARCHAR`/`VARBINARY` in `SQLGetTypeInfo` from 255 (<= 3.1.20) to 65535 in 3.1.22 ([ma_info.c](https://github.com/mariadb-corporation/mariadb-connector-odbc/blob/3.1.22/ma_info.c#L56)) and 65532 in 3.2.9, and `sqltype` used that value verbatim in `CREATE TABLE`. The package tests already fail this way on macOS (they use the 3.1.22 jll); Linux CI only passed because it downloads 3.1.20.

#401 fixes it in `sqltype` (treat `COLUMN_SIZE` as the maximum it is: above 8000 fall back to 255) and changes CI to also test a current connector, so the type-size fallback is covered by the existing `ODBC.load` test plus an explicit check.

On the terminator part: I could not reproduce it with Connector 3.2.9 (Homebrew, unixODBC) against MariaDB 13. `'abc'`, `''`, `'a\0'`, `'x\0\0'`, a 60k-character `TEXT` and `X'00'` all round-trip exactly in both fetch modes, and the driver reports the data length without the terminator. #401 adds those round trips as tests. If you still see a trailing NUL with 3.2.8, could you post the platform, driver manager, and the output of `DBInterface.execute(conn, "select 'abc' as s"; debug=true)` including `cursor.metadata`? A driver that reports the NUL inside the length would need a narrowly scoped workaround rather than stripping NULs from all strings (that would drop real data such as `"a\0"`).

**Comment by quinnj** [comment]
One more finding for anyone on Connector 3.2.x: with 3.2.9 (unixODBC, MariaDB 13) `SQLGetData` returns `22007 Invalid datetime format` for a NULL `TIME` value fetched through a prepared statement (direct `SQLExecDirect` is fine, 3.1.22 is fine). ODBC.jl `main` ignores that return code, so the cell silently receives the previous row's value instead of `missing`; #399 makes it an error. If you use prepared statements with nullable `TIME` columns on 3.2.x, check your results. I did not find an upstream report for it yet.

**Comment by hhaensel** [comment]
Thanks for the review. You are right, always stripping the NUL-byte is probably dangerous, it worked for my use case but will probably fail and no one knows why.
